tools/project.py

```diff
@@ -4,92 +4,97 @@
 sys.path.insert(0, ROOT)
 
 from shutil import move, rmtree
-from optparse import OptionParser
+from argparse import ArgumentParser
 from os import path
 
 from tools.paths import EXPORT_DIR, EXPORT_WORKSPACE, EXPORT_TMP
 from tools.paths import MBED_BASE, MBED_LIBRARIES
 from tools.export import export, setup_user_prj, EXPORTERS, mcu_ide_matrix
 from tools.utils import args_error, mkdir
 from tools.tests import TESTS, Test, TEST_MAP
+from tools.tests import test_known, test_name_known
 from tools.targets import TARGET_NAMES
 from tools.libraries import LIBRARIES
+from utils import argparse_lowercase_type, argparse_uppercase_type, argparse_filestring_type, argparse_many
 
-try:
-    import tools.private_settings as ps
-except:
-    ps = object()
 
 
 if __name__ == '__main__':
     # Parse Options
-    parser = OptionParser()
```

Comment:
There is a similar situation for `project.py` as I described above for `make.py`. Here are the valid combinations of arguments for `project.py` (the top level bullets are exclusive). cc @0xc0170:

- `-L <--list-tests>` - This should not require any other arguments to be set (ex. `-m`, `-i`, etc)
- `-S <--list-matrix>` - This should not require any other arguments to be set (ex. `-m`, `-i`, etc)
- `-E` - This should not require any other arguments to be set (ex. `-m`, `-i`, etc)
- One of the following commands (exclusively). All of the following should require all the export-related arguments (ex. `-m`, `-i`, etc):
  - Exactly one instance of `-n`
  - Exactly one instance of `-p`
  - At least one instance of `--source` (potentially more than one)